drivers/sdhc/sdhc_stm32.c

```diff
@@ -0,0 +1,659 @@
+			LOG_ERR("Failed to enable IO Function number %d\n", func);
+			return res;
+		}
+	}
+
+	HAL_SDIO_DirectCmd_TypeDef arg = {
+		.Reg_Addr = reg_addr, .ReadAfterWrite = raw_flag, .IOFunctionNbr = func};
+
+	if (direction == SDIO_IO_WRITE) {
+		uint8_t data_in = cmd->arg & SDIO_DIRECT_CMD_DATA_MASK;
+
+		res = HAL_SDIO_WriteDirect(&dev_data->hsd, &arg, data_in);
+	} else {
+		res = HAL_SDIO_ReadDirect(&dev_data->hsd, &arg, (uint8_t *)&cmd->response);
+	}
+
+	if (func != SDIO_FUNC_NUM_0) {
+		res = HAL_SDIO_DisableIOFunction(&dev_data->hsd, func);
+		if (res) {
+			LOG_ERR("Failed to disable IO Function number %d\n", func);
+			return res;
+		}
+	}
+
+	return res;
+}
+
+static int sdhc_stm32_rw_extended(struct sdhc_stm32_data *dev_data, struct sdhc_command *cmd,
+				  struct sdhc_data *data)
+{
+	int res;
+	bool direction = (cmd->arg >> SDIO_CMD_ARG_RW_SHIFT);
+	bool increment = (cmd->arg & BIT(SDIO_EXTEND_CMD_ARG_OP_CODE_SHIFT));
+	bool is_block_mode = (cmd->arg & BIT(SDIO_EXTEND_CMD_ARG_BLK_SHIFT));
+	uint8_t func = (cmd->arg >> SDIO_CMD_ARG_FUNC_NUM_SHIFT) & 0x7;
+	uint32_t reg_addr = (cmd->arg >> SDIO_CMD_ARG_REG_ADDR_SHIFT) & SDIO_CMD_ARG_REG_ADDR_MASK;
+
+	if (func != SDIO_FUNC_NUM_0) {
+		res = HAL_SDIO_EnableIOFunction(&dev_data->hsd, func);
+		if (res) {
+			LOG_ERR("Failed to enable IO Function number %d\n", func);
+			return res;
+		}
+	}
+
+	HAL_SDIO_ExtendedCmd_TypeDef arg = {.Reg_Addr = reg_addr,
+					    .IOFunctionNbr = func,
+					    .Block_Mode = is_block_mode ? SDMMC_SDIO_MODE_BLOCK
+									: HAL_SDIO_MODE_BYTE,
+					    .OpCode = increment};
+
+	if (is_block_mode && (func != SDIO_FUNC_NUM_0)) {
+		res = HAL_SDIO_SetBlockSize(&dev_data->hsd, func, data->block_size);
+		if (res != HAL_OK) {
+			LOG_ERR("Failed to change block size.\n");
+			return res;
+		}
+	}
+
+	if (direction == SDIO_IO_WRITE) {
+		if (data->data) {
+			res = HAL_SDIO_WriteExtended(&dev_data->hsd, &arg, data->data,
+						     ((data->blocks) * (data->block_size)),
+						     data->timeout_ms);
+		} else {
+			LOG_ERR("Invalid data buffer passed to CMD53 (write).");
+			res = -EINVAL;
+		}
+	} else {
+		res = HAL_SDIO_ReadExtended(&dev_data->hsd, &arg, (uint8_t *)&cmd->response,
+					    ((data->blocks) * (data->block_size)),
+					    data->timeout_ms);
+	}
+
+	if (func != SDIO_FUNC_NUM_0) {
+		res = HAL_SDIO_DisableIOFunction(&dev_data->hsd, func);
+		if (res) {
+			LOG_ERR("Failed to disable IO Function number %d\n", func);
+			return res;
+		}
+	}
+
+	return res;
+}
+
+static int sdhc_stm32_switch_to_1_8v(const struct device *dev)
+{
+	uint32_t res = 0;
+	uint32_t resp_val;
+	struct sdhc_stm32_data *data = dev->data;
+
+	/* Check if host supports 1.8V signaling */
+	if (!data->props.host_caps.vol_180_support) {
+		LOG_ERR("Host does not support 1.8V signaling");
+		return -ENOTSUP;
+	}
+
+	/* Ask the card if it supports 1.8V signaling (CMD5 with S18R) */
+	res = SDMMC_CmdSendOperationcondition(data->hsd.Instance, SDIO_OCR_SDIO_S18R, &resp_val);
```

Comment:
This is performed within the SD stack- it is not the responsibility of the driver and should not be implemented here.